### PR TITLE
Select samples grain size

### PR DESCRIPTION
**Pull request by ShawnPatrick-Barhorst**
This is my function for sam to look at

### CODE REVIEW

File: select_samples_grain_size.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+from scipy.ndimage import center_of_mass
+
+
+def select_samples_grain_size(mask, bins, num_samples=1, setting='highest'):
```
Comment (swelborn) [inline select_samples_grain_size.py]: Here we can use type annotations. For example:

```python
def select_samples_grain_size(mask: np.ndarray, bins: int, num_samples: int = 1, setting: str = "highest") -> list[list[int]]:
```

Further, we want to use an enum for the setting (there are only two options!):

so --
 
```python
from enum import Enum
...

class GrainSizeSetting(str, Enum):
	HIGHEST = "highest"
	LOWEST = "lowest"

# and in the def:
 ... setting: GrainSizeSetting = GrainSizeSetting.HIGHEST):
 
 ...
 
 if setting == GrainSizeSetting.HIGHEST:
 	...
 elif setting == GrainSizeSetting.LOWEST:
 	...
 else:
 	...
 ```

File: select_samples_grain_size.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    for ii in range(bins):
+        bounds = (bin_boundaries[ii], bin_boundaries[ii+1])
+        bool_arr = np.array([sizes_new>=bounds[0], sizes_new<bounds[1]])
+        size_args.append(list(np.argwhere(bool_arr.all(0)).T[0]))
+
+    # Take samples from each bin
+    sample_list = []
+
+    for ii in range(bins):
```
Comment (swelborn) [inline select_samples_grain_size.py]: In general, we should be using useful loop variables. 

for example (I am not sure this actually means `bin_num`, so make it what it means):

```python
for bin_num in range(bins):
	...
```

File: select_samples_grain_size.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    # Find the corresponding indices
+    size_args = []
+
+    for ii in range(bins):
+        bounds = (bin_boundaries[ii], bin_boundaries[ii+1])
+        bool_arr = np.array([sizes_new>=bounds[0], sizes_new<bounds[1]])
+        size_args.append(list(np.argwhere(bool_arr.all(0)).T[0]))
+
+    # Take samples from each bin
+    sample_list = []
```
Comment (swelborn) [inline select_samples_grain_size.py]: What type is sample_list?

For example:

```python
sample_list: list[np.ndarray] = []
```

or if you are converting it to another list,

```python
sample_list: list[list[int]] = []
```

File: select_samples_grain_size.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    if setting == 'highest':
+        ii = 0
+    if setting == 'lowest':
+        ii = bins-1
+       
+    sl = sample_list[ii]
+
+    coord_ii = []
+    for jj in range(len(sl)):
```
Comment (swelborn) [inline select_samples_grain_size.py]: Change loop variable to something useful:

```python
for sample_num in range(len(sl)):
	...
```

File: select_samples_grain_size.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    # Remove the zeroth element (the background)
+    sizes_new = sizes[1:]
+
+    # Bin these clusters by size
+    hist = np.histogram(sizes_new, bins=bins)
+    bin_boundaries = hist[1]
+
+    # Find the corresponding indices
+    size_args = []
```
Comment (swelborn) [inline select_samples_grain_size.py]: what type is size_args?

File: select_samples_grain_size.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    for ii in range(bins):
```
Comment (swelborn) [inline select_samples_grain_size.py]: change to useful loop variable

**Review by swelborn — CHANGES_REQUESTED** [review]
Looking good. Inside the function we should specify types, and be more descriptive with loop variables. This helps others figure out what things mean in your code.

Always add typing. There are 1,000 benefits to doing this, including self-documentation, and your IDE (e.g., vscode) can pick up on what you are doing wrong and right if you make a small change.

**Review by pmiam — COMMENTED** [review]
When num_samples is increased, the returned centroid-derived coordinates are still repetitive.

My understanding is that the point of the random sampling is to drive exploration of the sample. Probably a minor detail at this stage.

**Review by swelborn — APPROVED** [review]

**Pull request merged**